### ISSUE TITLE
Potential improvements for the Deep Learning with Flux.jl course on JuliaAcademy.com

### ISSUE THREAD

**Issue by logankilpatrick** [JuliaComputing/JuliaAcademyMaterials#26]
Thread created during Google Code-In.

**Comment by mikolajhojda** [comment]
In the course should be review test. I think it will help in learning. We could remember informations easier. The beginning of the course was chaotic for me. Maybe it should be how to install it etc.

**Comment by chinglamchoi** [comment]
1. General Things:
- Can integrate quizzes; videos can be broken down into separte, shorter ones; amazing plots and visualisation (esp the contour plots and visualising decision boundaries for multi-dimensional data)

2. Handwritten & Built-in Functions 
Really liked how he included both handwritten and built-in methods of building aspects of the model (e.g. custom-building activation functions, optimisers, etc). The handwritten functions tutorial is useful in understanding the mechanisms behind each Flux built-in function, and also comes in handy when you have to try to define you own custom functions for your architecture (mobilenet's relu6 activation comes to mind).

3. Very Well-Explained
Amazing how he used linear transformation to explain how model complexity is aggregated (using multilpe lienar layers doesn't make a complex model at all). Also, I really appreciate how he elaborated on which loss function to choose and why, such as through plotting MSE & Cross Entropy and visualising their domains. It's also great that he decompiled the model architecture, showing us the code and logic behind each function. 

4. Very Easy to Follow & Recreate: 
A lot of tutorials skip right to defining the model architecture & training process and omit the data loading (creating one-hot labels, batches, etc) and visualisation. Knowledge taught in this course fully equips us to use Flux on our own custom datasets. 

5. More on GPU Acceleration
Perhaps more could have been said about GPU acceleration and how to leverage the power of CUDA. It would also be nice to cover CuArrays and GPU VRAM memory allocation, as it seems like OutofMemory errors are common issues when training with GPUs.

6 A Bit More on Choosing Hyperparameters & Optimisers
- Adam versus SGD versus AMSGrad, and stability problems
- effect of learning rate on convergence and learning weight decay
- effect of batch size on generalisation, memory, speed (Yann LeCun: "Friends dont let friends use minibatches larger than 32.")
- I think the instructor could have expatiated on hyperparameter tuning

7. Some Advanced Topics
- It would have been nice to go over how one could create skip connections in Flux. While it is quite challenging to do so, many baseline (Resnet) and state-of-the-art methods (Efficient-Net, based on Mobilenet v2) utilise skip connections for accuracy gains with deep models, making skip connections an extremely relevant topic in ML with Flux.
- Would have been nice to cover transfer learning, how to load pre-trained weights on Imagenet for finetuning on custom datasets. This is a widely used and very potent technique, I'm sure it would have been useful to students!

**Comment by CompSciDZ** [comment]
For me, this was where neural networking and machine learning got confusing, but replaying the videos and watching them over again is helpful.

For this course, due to the nature of complexity and the pace, I think prerequisites should be set and defined. Additionally, to truly understand these concepts, I think that the videos need to be rewatched and that everyone should look through the notebook. The instructor goes beyond their ability to clarify things, but additional clarification may be necessary in some areas.

In this course, modeling is necessary. Although we focus on the code, modeling what each statement does is critical to grasp these concepts. I think more visual representations need to be incorporated to help students understand.

As far as the code goes, I think that this course threw a lot of code at you and is hard to follow through. Courses prior to this did not go much into the code (Intro to Julia did not go to Machine Learning, Foundations of Machine Learning was mainly modeling, and Knet did not go into this complexity of the code). 

This course is something that you do not comprehend the first time, but with the instructor's explanations, it helps make it easier, however, it is highly recommended you have a strong background in understanding machine learning to fully experience the course.

**Comment by PseudoCodeNerd** [comment]
#### Some background before this review. 
I already had previous experience with Machine Learning through Andrew Ng's Coursera Course and books that my brother recommended. I also had the understanding of the math behind many of the concepts discussed in the course. So, I didn't find the course too intimidating instead, it proved to be a great opportunity to learn Flux
So now let's get to the main thing.

I really liked the way of teaching. No nonsense and clear. The instructor very well explained  all the concepts and the notebooks provided were very high quality too. They were extensively documented and were a perfect complement to the lectures. 

As many have previously mentioned, I'd agree that the course is indeed of a higher difficulty (It might also be difficult to understand as those who're writing a reviews are high-school students .themselves).

The course organisation is really good. Starting from Flux and all the way to making our own deep NN.  The instructor deserves appreciation for the explanation he tried to provide for nearly everything he did and the way I was gripped till the end of every lecture. I'd recommend a quiz question every 10 minutes or so inside the video itself (MCQ). Something in the Stanford ML MOOC that intrigued and help me clear concepts. 
 
As for the platform, I'd suggest adding captions to incorporate a larger learning audience from the world and also to address times when the voice isn't comprehensible (only some times though). The course should also indicate the prerequisites like some understanding of probability theory and linear algebra and calculus which could be helpful in understanding how things are actually working.  

It is indeed an awesome course but for the right audience.
👍 Keep It Up Guys !

**Comment by TheComputerM** [comment]
It was a great tutorial, better than Flux's documentation and other tutorials available on YouTube. 

Even a beginner with no prior knowledge of machine learning would be able to follow along. The explanations provided for Neural Networks and loss functions were accurate and easy to understand.

Again as with every other tutorial on Julia academy the problem comes from lack of exercises and being able to interact with sample notebooks along the way. I felt that this tutorial was __short__ as it could demonstrate so much more with flux and machine learning. And an in-depth analysis of different machine learning models such as CNN, RNN and LSTM was missing.

Please take a look at https://mybinder.org/ to host the sample jupyter notebooks.

**Comment by ishita272001** [comment]
Although the course was very good.
But it needs some improvements -
1.The course was quite difficult to understand as a beginner.
2.As every course I should tell some tests at last should be  added.
tests can added.Questions can be some handwriting is given and tells to identify it with neural network.And more................
3.Some interaction and sound quality of video should be increased.
4.The PDF contain code but not in an organized way.
5.As the course has only four parts with much time.
The course should be divided in much smaller parts.Like each part is 3-8 mins.
HOPE YOU MAKE IMPROVEMENTS.
THANKS

**Comment by Redeem-Grimm-Satoshi** [comment]
Never used flux before but this course helped me understand the basics of neural networks with flux and how to use it to recognize handwriting. I was satisfied with the video length and explanation. I think if the package flux.jl was manually installed on JuliaPro or Julia so users can see how it is being installed before using it. Review questions too can be provided to remember what has been taught. Lastly,  I also want to congratulate the mentor for doing a great job!

**Comment by josobar** [comment]
I have only started using Julia when GCI 2019/2020 began and I had little experience of machine learning prior to this tutorial. This course builds off of the previous one very well and simply translates its concepts into Julia code. The instructor is clearly credible and the video and audio quality is good. 

Suggestions:
 - The course sections are rather fast-paced and I had to watch almost all of them at least two times to understand what the instructor was talking about.
- The course includes a lot of code in Jupyter Notebook and its a lot to take in. I found that this is helped by me running the same code in a separate window and messing with a little bit to see how it works. 
- This is another Julia ACademy course that only offers videos and lacks other forms of interactivity, but this can be fixed wiht quizzes and other possible GCI tasks.

**Comment by fungktt** [comment]
Hi everyone! I just wanted to add my thoughts to this course over here.

Coming from someone who has no prior experience with ML, I must say that this tutorial is very intuitive. I did not know any calculus and I soon realised that I didn't really need.

It would be great if we could have more interactivity with the content to reinforce the knowledge covered in this course with something to do with testing: such as quizzes or some problems.

I must agree that some of the stuff towards the first video was quite poor and I was personally slightly confused about the sigmoid. It would be great if there could be more of a linkage between the Foundations of Machine Learning course and this.

I do enjoy the handwritten aspects of the learning models very interesting but I feel like it was emphasised a bit too much and a large part of the video was covering that which defeated the whole purpose of having Flux to simplify the machine learning process. 

At the same time, I do hope that there could be more on CUDA, optimisers and different loss functions such as crossentropy and MSE. Although we were introduced to both, it could be better if there was more explanation as to why users should use a certain one in a certain situation.

In addition, I really like the image classification aspect of the course in the last video but I wish there was also a tutorial for non-Grayscale images to fill the whole picture.

Also, I would recommend the Atom/Juno setup for this course because I had a bad experience with JuliaPro while doing this course and some of the code was not compatible on JuliaPro, which was introduced to all the users on the "Introduction to JuliaAcademy.com" course.

Thanks :)

**Comment by eahmed123** [comment]
I am done with the course but they are not giving me the certificate. I don't know what is the problem. In this course there are two problems. On 
![JuliaAcademy (1)](https://user-images.githubusercontent.com/59716131/72304266-1071ff00-363e-11ea-90c6-f7114b3b8f77.png)
is that I have done course but no certificate. Secondly, there is not even one quiz to test the knowledge of the student.
Let me give the example of question how it should be....
Question 1: Julia has a powerful package that does much of a heavy lifting for us called "Flux.jl"?
A. True
B. False
There should be specific time given to the student to complete question or the whole quiz.

**Comment by Ninad10code** [comment]
At the beginning of the course, there must be a video describing what is deep learning and what is the difference between deep learning and machine learning cause at the beginning I wasn't able to know the difference between the two or how the two are related.
The best part of the course understanding was the data visualization of all the inputs given to training the model. Understanding the non-linearity curve was a bit difficult to imagine how thee curve gets generated,
Recognizing handwriting with neural network was a very interesting part of the course
Thank you.

**Comment by dhruv2604-create** [comment]
Hi people! I believe that JuliaAcademy is an excellent source for not only experts but beginners too. It has a course for everyone. It was during GCI 2019-2020 that I came to know about Julia and by completing "Introduction to Julia" and "Deep Learning with Flux.jl" courses I have become quite familiar with Julia. Although I would like to recommend mini-tests or a final test to test a user's knowledge and how much has the course helped him or her.

**Comment by rajat-png** [comment]
FEEDBACK : 
1. Tutorial is good and best but there is no chat system for students doubt. It seems inserting the chat system (direct talk to trainer) the all students are clearing the doubt from trainer. "I HIGHLY RECOMMENDED THIS"
2. I've successful completed the Deep Learning with Flux.jl course on your JuliaAcademy.org but there is no lesson to practice, when insert a practice lesson than students after learn the course, live do practice and really completed the course.
3. Made practice set for students.
THANKYOU

**Comment by king398** [comment]
Wish there could be a quiz. Also if they partnered with TensorFlow for Julia instead of flux cause many students use it. Also the videos could be broken into chapters like udacity. Overall the course was good

**Comment by itqorst03** [comment]
Here's my review of the Deep Learning with Flux.jl course:

I found the course a bit tough and had to watch some videos 2-3 times to understand. The videos were a bit fast paced and it took some time for me to understand what the tutor was saying. It would be better if there were subtitles. The course should be more interactive. There should be more quizzes. Also, I think the audio quality could have been better. 30 minute videos are a bit too long. It would be better if they were broken down into several 8-10 minute videos.
I would rate this course 7/10

**Comment by king398** [comment]
1. the course cold have shown prerequisites before.

2.  the chapter should have been broken into shorter videos or chapters like udacity

3. if there was a summary it would be helpful

4. a quiz after each chapter would have good

5. i give the rating to the course of 6/10

6.  this is not an important thing but if you could partner with tensor flow it would be very good

**Comment by Akshat-mehrotra** [comment]
*I'll try not to suggest things people have already suggested.

Only after watching the video series by 3 blue one brown on neural networks and machine learning, I had a respectable benchmark. 
This course lacks a detailed and satisfactory explanation of the core concepts of ML. 3B1B explains most of the concepts without even touching programming. I understand that his is more of a math-oriented video series, however, the whole field of ML is heavily math-oriented. Emphasize on math should undoubtedly increase. One of the ways to implement this so that people who like and dislike math could be satisfied could be to create side videos purely focused on math only for those who are looking for a mathematical approach.
Otherwise, the course is great for someone who is familiar with the concepts of ML and wants a quick introduction to it in Julia.

**Comment by D3nii** [comment]
While taking the course, I was like why are there this big videos, it gets pretty boring tbh. There must be short videos on labelled topics and also some kind of tests like DataCamp has in its courses. I have taken 3-4 courses from this website and need tests but can't have them, so, my advice is to get the tests part done asap. Otherwise, the course is okay and is a nice pathway for students to follow.

**Comment by kgauld1** [comment]
This course was super interesting and a great way to learn about machine learning in Julia.  The explanations were thorough and the material was extremely interesting.  I really enjoyed taking this course, but it might be improved by breaking up longer videos into a few smaller videos.

**Comment by amanjincodeout** [comment]
Truly speaking about the course that it is not only encouraging but bit it is interesting
The educator is very nice although i am disappointed because inbetween videos there r no tests 
so student can't understand that he learnt correctly or not
but although in short videos it has knowledgeable.

**Comment by salemalem** [comment]
Hello there everybody! I fair needed to include my considerations to this course over here. Coming from somebody who has no earlier involvement with ML, I must say that this instructional exercise is exceptionally natural. I did not know any calculus and was by and by freaking out but I before long figured it out that I didn't truly require any of it.

**Comment by amanjincodeout** [comment]
I am very happy to learn Julia Computing through this course :
1] Lectures are very clear and useful to understand the difficult contents of the module. Interesting
module.As I have seen that difficult topics are covered,but it is not in depth but it gives us basic idea about that.
Although the module content is very difficult for me when I read the books,and to overcome this,this course gives me kick.

2]The professor has a good speaking and teaching style which keeps me interested. Lots of tough  examples which make it easier to understand. Very good slides which are well formulated,
but I do not like presentation, as there should be relations between topic and real world life.
As I have seen and read in my books with me.

3]One most important thing that i do not like about this course that at starting there was only one video i.e around 3 min , that i don't like,as because introduction is key basic according to me,and that should cover all the overview, which was not covered

4]At last, that it should have test in between videos like codeacademy,but it was not there.
Also there should be any platform to clear doubts,as many other sites including this has not given.

**Comment by harshpoddar03** [comment]
I had a wonderful experience doing this course
the professor explained everything carefully and which made the learning experience more interacting 
.
But the course was very long and time-consuming. I found many things confusing and the professor was a little unclear something. I was not able to understand his handwriting which also caused confusion

But overall, it was a good way of learning new things.

**Comment by salemalem** [comment]
Hi! I think the course can be **better** with:

1.  Practice. I couldn't see any practice exercises as Udemy courses can offer.
2. There is any interaction or quizzes at the end of the lecture as in Udemy. It will be very helpful to consolidate knowledge.
3. There is no source code used in the lesson. I want to download source code or just take a look, but, unfortunately, I don't have this opportunity.

Generally, the course is fine.

**Comment by AlexmMark** [comment]
Hello! Alex M from Google Code-In
Some things about this course;

1. I think it would be nice if subtitles were included in the videos! It would be easier to understand everything the instructor says. Also, it would help anyone that owns a device that has trouble with sound.
2. In the notebook tutorial, a link leading the viewer to the online jubyter demo notebook would be usefull.
3. As a student familiar with java, I found the language to be great and intuitive!
4. I had a question about the pop! function in Dictionaries which did not get answered, why do we get the data the deleted element points to and not the element itself too? Will watch again, but I did have this question!
5. I did not fully comprehend the significance of the second parameter variable in Array{}
6. I really love the fact that there is text explanation of the code, it helped me follow the tutorial and stay focused to what I was seeing and understanding.
7. Some excercises between the tutorials would be great!
8. And lastly, each time the tutorial video was completed and I was automatically taken to the next one, I found it to be a bit annoying that I had to readjust the screen to properly view the video. I was using Firefox web browser.

Overall, I found the language very intuitive for someone who has occupied themselves in the past with programming, but also it could be a beginner tutorial if there were a bit more exercises!
That's all from me!
